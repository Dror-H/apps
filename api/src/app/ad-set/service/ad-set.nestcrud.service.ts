import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { AdSet } from '../data/ad-set.entity';
import { AdSetRepository } from '../data/ad-set.repository';

@Injectable()
export class AdSetNestCrudService extends TypeOrmCrudService<AdSet> {
  constructor(@Inject(AdSetRepository) repo: AdSetRepository) {
    super(repo);
  }
}
