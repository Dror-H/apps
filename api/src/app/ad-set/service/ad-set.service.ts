import {
  AdSetCreationDTO,
  WorkspaceDTO,
  AdSetDTO,
  ResponseSuccess,
  SupportedProviders,
  TargetingDto,
  CacheTtlSeconds,
} from '@instigo-app/data-transfer-object';
import { ThirdPartyAdSetApiService } from '@instigo-app/third-party-connector';
import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { Cache } from 'cache-manager';
import { groupBy } from 'lodash';
import { from, Observable } from 'rxjs';
import { map, mergeMap, switchMap } from 'rxjs/operators';
import { AdSet } from '../data/ad-set.entity';
import { AdSetRepository } from '../data/ad-set.repository';

@Injectable()
export class AdSetService {
  @Inject(ThirdPartyAdSetApiService)
  private readonly thirdPartyAdSetApiService: ThirdPartyAdSetApiService;

  @Inject(AdSetRepository)
  private readonly repo: AdSetRepository;

  @Inject(CACHE_MANAGER)
  private readonly cacheManager: Cache;

  async create(adSet: Partial<AdSetCreationDTO>, currentWorkspace: WorkspaceDTO): Promise<AdSetDTO> {
    const { provider, adAccount, campaign } = adSet;
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const builtAdSet = await this.thirdPartyAdSetApiService.create({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      adSet: adSet,
    });
    builtAdSet.adAccount = adAccount;
    builtAdSet.campaign = campaign;
    return this.repo.save(builtAdSet);
  }

  bulkPatch(options: { adSets: Partial<AdSetDTO>[]; workspace: WorkspaceDTO }): Observable<any> {
    const { adSets, workspace } = options;
    void this.cacheManager.reset();
    return from(Object.entries(groupBy(adSets, 'provider'))).pipe(
      mergeMap(([provider, entries]) => {
        const { accessToken } = workspace.owner.getAccessToken({ provider });
        return this.thirdPartyAdSetApiService.bulkPatch({ provider, accessToken, adSets: entries });
      }),
      switchMap(() => this.repo.save(adSets)),
      map(() => new ResponseSuccess('AdSets successfully updated')),
    );
  }

  async findAdSetTargeting(options: {
    workspace: WorkspaceDTO;
    provider: SupportedProviders;
    adSetId: string;
  }): Promise<TargetingDto> {
    const { adSetId, provider, workspace } = options;
    const cacheKey = `adSetTargeting-${provider}-${adSetId}`;
    let result = await this.cacheManager.get(cacheKey);
    if (result) {
      return result as any;
    }
    const { accessToken } = workspace.owner.getAccessToken({ provider });
    result = await this.thirdPartyAdSetApiService.findAdSetTargeting({
      accessToken,
      adSetId,
      provider,
    });
    await this.cacheManager.set(cacheKey, result, { ttl: CacheTtlSeconds.ONE_HOUR });
    return result as any;
  }
}
