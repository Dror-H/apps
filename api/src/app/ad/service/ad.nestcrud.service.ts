import { Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { Ad } from '../data/ad.entity';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';

@Injectable()
export class AdNestCrudService extends TypeOrmCrudService<Ad> {
  constructor(
    @InjectRepository(Ad)
    public repo: Repository<Ad>,
  ) {
    super(repo);
  }
}
