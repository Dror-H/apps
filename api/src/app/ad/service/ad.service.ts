import { CACHE_MANAGER, Inject, Injectable } from '@nestjs/common';
import { AdCreationDTO, AdDTO, AdUpdateDTO, ResponseSuccess, WorkspaceDTO } from '@instigo-app/data-transfer-object';
import { Ad } from '@api/ad/data/ad.entity';
import { ThirdPartyAdApiService } from '@instigo-app/third-party-connector';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Cache } from 'cache-manager';
import { from, Observable } from 'rxjs';
import { groupBy } from 'lodash';
import { map, mergeMap, switchMap } from 'rxjs/operators';

@Injectable()
export class AdService {
  @Inject(ThirdPartyAdApiService)
  private readonly thirdPartyAdApiService: ThirdPartyAdApiService;

  @InjectRepository(Ad)
  private readonly adRepository: Repository<Ad>;

  @Inject(CACHE_MANAGER)
  private readonly cacheManager: Cache;

  async create(ad: Partial<AdCreationDTO<any>>, currentWorkspace: WorkspaceDTO): Promise<AdDTO> {
    const { provider, adAccount } = ad;
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const builtAd = await this.thirdPartyAdApiService.create({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      ad,
    });
    builtAd.adAccount = adAccount;
    return this.adRepository.save(builtAd);
  }

  async bulkCreate(ads: Partial<AdCreationDTO<any>>[], currentWorkspace: WorkspaceDTO): Promise<Ad[]> {
    const { provider, adAccount, campaign } = ads[0];
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider });
    const createdAds = await this.thirdPartyAdApiService.createMany({
      accessToken,
      adAccountProviderId: adAccount.providerId,
      provider,
      ads: ads,
    });
    const adSet = ads[0].providerSpecificFields?.adSet;
    const mappedAds = createdAds.map((item) => ({ ...item, adAccount, campaign, adSet }));
    return this.adRepository.save(mappedAds);
  }

  public async updateOne(ad: AdUpdateDTO, currentWorkspace: WorkspaceDTO): Promise<Ad> {
    await this.cacheManager.reset();
    const { accessToken } = currentWorkspace.owner.getAccessToken({ provider: ad.provider });
    await this.thirdPartyAdApiService.updateAd({ provider: ad.provider, accessToken, ad });
    const newAd = {
      id: ad.id,
      name: ad.name,
      objectStorySpec: { ...ad.data, adTemplateType: ad.adTemplateType },
    } as Ad;
    return await this.adRepository.save(newAd);
  }

  bulkPatch(options: { ads: Partial<AdDTO>[]; workspace: WorkspaceDTO }): Observable<any> {
    const { ads, workspace } = options;
    void this.cacheManager.reset();
    return from(Object.entries(groupBy(ads, 'provider'))).pipe(
      mergeMap(([provider, entries]) => {
        const { accessToken } = workspace.owner.getAccessToken({ provider });
        return this.thirdPartyAdApiService.bulkPatch({ provider, accessToken, ads: entries });
      }),
      switchMap(() => this.adRepository.save(ads)),
      map(() => new ResponseSuccess('Ads successfully updated')),
    );
  }
}
