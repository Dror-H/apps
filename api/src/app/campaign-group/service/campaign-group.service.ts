import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { CampaignGroup } from '../data/campaign-group.entity';
import { CampaignGroupRepository } from '../data/campaign-group.repository';

@Injectable()
export class CampaignGroupNestCrudService extends TypeOrmCrudService<CampaignGroup> {
  constructor(@Inject(CampaignGroupRepository) public repo: CampaignGroupRepository) {
    super(repo);
  }
}
