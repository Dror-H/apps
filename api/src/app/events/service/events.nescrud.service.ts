import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { EventsRepository } from '@api/events/data/events.repository';
import { EventsEntity } from '@api/events/data/events.entity';

@Injectable()
export class EventsNestCrudService extends TypeOrmCrudService<EventsEntity> {
  constructor(@Inject(EventsRepository) eventsRepository: EventsRepository) {
    super(eventsRepository);
  }
}
