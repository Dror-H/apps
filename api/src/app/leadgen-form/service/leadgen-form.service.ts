import { Inject, Injectable } from '@nestjs/common';
import { TypeOrmCrudService } from '@nestjsx/crud-typeorm';
import { LeadgenForm } from '../data/leadgen-form.entity';
import { LeadgenFormRepository } from '../data/leadgen_form.repository';

@Injectable()
export class LeadgenFormNestCrudService extends TypeOrmCrudService<LeadgenForm> {
  constructor(@Inject(LeadgenFormRepository) public repo: LeadgenFormRepository) {
    super(repo);
  }
}
