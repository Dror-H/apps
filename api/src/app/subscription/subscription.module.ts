import { CachingModule } from '@api/database';
import { InsightsServicesModule } from '@api/insights/services/insights-services.module';
import { UserModule } from '@api/user/user.module';
import { StripeModule } from '@golevelup/nestjs-stripe';
import { applyRawBodyOnlyTo } from '@golevelup/nestjs-webhooks';
import { forwardRef, MiddlewareConsumer, Module, RequestMethod } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { PaymentController } from './controllers/payment.controller';
import { SubscriptionCronJobsController } from './controllers/subscription-cron-jobs.controller';
import { SubscriptionController } from './controllers/subscription.controller';
import { InvoiceService } from './services/invoice.service';
import { PaymentService } from './services/payment.service';
import { SubscriptionCronJobService } from './services/subscription-cron-jobs.service';
import { SubscriptionService } from './services/subscription.service';

@Module({
  imports: [
    forwardRef(() => UserModule),
    InsightsServicesModule,
    StripeModule.forRootAsync(StripeModule, {
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: (configService: ConfigService) => ({
        apiKey: configService.get<string>('STRIPE_SECRET_KEY'),
        apiVersion: '2020-08-27',
        typescript: true,
        webhookConfig: {
          stripeWebhookSecret: configService.get<string>('STRIPE_WEBHOOK_SECRET'),
        },
      }),
    }),
    CachingModule,
  ],
  providers: [SubscriptionService, InvoiceService, PaymentService, SubscriptionCronJobService],
  controllers: [SubscriptionController, PaymentController, SubscriptionCronJobsController],
  exports: [SubscriptionService],
})
export class SubscriptionModule {
  configure(consumer: MiddlewareConsumer): void {
    applyRawBodyOnlyTo(consumer, {
      method: RequestMethod.ALL,
      path: 'stripe/webhook',
    });
  }
}
