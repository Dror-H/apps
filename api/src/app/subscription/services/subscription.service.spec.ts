import { StripeModule } from '@golevelup/nestjs-stripe';
import { MockRepository, userMock } from '@instigo-app/api-shared';
import { Test } from '@nestjs/testing';
import { Repository } from 'typeorm';
import { SubscriptionService } from './subscription.service';
const stripeSubscription = {
  id: 'sub_1Jf7BWKlYpB2IMP7vWM2pUOh',
  object: 'subscription',
  application_fee_percent: null,
  automatic_tax: {
    enabled: false,
  },
  billing_cycle_anchor: 1633039200,
  billing_thresholds: null,
  cancel_at: null,
  cancel_at_period_end: false,
  canceled_at: null,
  collection_method: 'charge_automatically',
  created: 1632938834,
  current_period_end: 1633039200,
  current_period_start: 1632938834,
  customer: 'cus_KJkeW5VmQDpYCA',
  days_until_due: null,
  default_payment_method: null,
  default_source: null,
  default_tax_rates: [],
  discount: null,
  ended_at: null,
  items: {
    object: 'list',
    data: [
      {
        id: 'si_KJkgnL0Cs6l6UM',
        object: 'subscription_item',
        billing_thresholds: null,
        created: 1632938834,
        metadata: {},
        plan: {
          id: 'price_1JewjrKlYpB2IMP7GTT1OKMv',
          object: 'plan',
          active: true,
          aggregate_usage: null,
          amount: null,
          amount_decimal: null,
          billing_scheme: 'tiered',
          created: 1632898679,
          currency: 'usd',
          interval: 'month',
          interval_count: 1,
          livemode: false,
          metadata: {},
          nickname: null,
          product: 'prod_KGIb3Cm3ThJEcd',
          tiers_mode: 'volume',
          transform_usage: null,
          trial_period_days: null,
          usage_type: 'licensed',
        },
        price: {
          id: 'price_1JewjrKlYpB2IMP7GTT1OKMv',
          object: 'price',
          active: true,
          billing_scheme: 'tiered',
          created: 1632898679,
          currency: 'usd',
          livemode: false,
          lookup_key: null,
          metadata: {},
          nickname: null,
          product: 'prod_KGIb3Cm3ThJEcd',
          recurring: {
            aggregate_usage: null,
            interval: 'month',
            interval_count: 1,
            trial_period_days: null,
            usage_type: 'licensed',
          },
          tax_behavior: 'unspecified',
          tiers_mode: 'volume',
          transform_quantity: null,
          type: 'recurring',
          unit_amount: null,
          unit_amount_decimal: null,
        },
        quantity: 1,
        subscription: 'sub_1Jf7BWKlYpB2IMP7vWM2pUOh',
        tax_rates: [],
      },
    ],
    has_more: false,
    total_count: 1,
    url: '/v1/subscription_items?subscription=sub_1Jf7BWKlYpB2IMP7vWM2pUOh',
  },
  latest_invoice: 'in_1Jf7BWKlYpB2IMP708aJ6Eyd',
  livemode: false,
  metadata: {
    user: 'lupu60@gmail.com',
    workspace: '336f821a-9217-46bb-9e7a-e7ca9e8f3036',
  },
  next_pending_invoice_item_invoice: null,
  pause_collection: null,
  payment_settings: {
    payment_method_options: null,
    payment_method_types: null,
  },
  pending_invoice_item_interval: null,
  pending_setup_intent: null,
  pending_update: null,
  plan: {
    id: 'price_1JewjrKlYpB2IMP7GTT1OKMv',
    object: 'plan',
    active: true,
    aggregate_usage: null,
    amount: null,
    amount_decimal: null,
    billing_scheme: 'tiered',
    created: 1632898679,
    currency: 'usd',
    interval: 'month',
    interval_count: 1,
    livemode: false,
    metadata: {},
    nickname: null,
    product: 'prod_KGIb3Cm3ThJEcd',
    tiers_mode: 'volume',
    transform_usage: null,
    trial_period_days: null,
    usage_type: 'licensed',
  },
  quantity: 1,
  schedule: null,
  start_date: 1632938834,
  status: 'active',
  transfer_data: null,
  trial_end: null,
  trial_start: null,
};
describe('Subscription Test suite', () => {
  let service: SubscriptionService;
  let userRepository: Repository<any>;
  beforeEach(async () => {
    const module = await Test.createTestingModule({
      imports: [
        StripeModule.forRoot(StripeModule, {
          apiKey: '123',
        }),
      ],
      providers: [
        { provide: 'Connection', useValue: new MockRepository() },
        { provide: 'SubscriptionRepository', useValue: new MockRepository() },
        { provide: 'MailerService', useValue: { sendMail: jest.fn().mockResolvedValue({}) } },
        { provide: 'AdSpendService', useValue: { userAdSpend: jest.fn().mockResolvedValue({ total: 100 }) } },
        { provide: 'UserRepository', useValue: new MockRepository() },
        SubscriptionService,
      ],
    }).compile();

    service = module.get<SubscriptionService>(SubscriptionService);
    userRepository = module.get<Repository<any>>('UserRepository');
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should create customer', async () => {
    // Arrange
    const customer = {
      id: 'cus_KGjEQgjxBxlXpa',
      object: 'customer',
      name: 'Bogdan Lupu',
    };
    const user = userMock();
    userRepository.findOne = jest.fn().mockResolvedValue(user);
    userRepository.update = jest.fn().mockResolvedValue(user);
    (service as any).stripeService.customers.list = jest.fn().mockResolvedValue({ data: [] });
    (service as any).stripeService.customers.create = jest.fn().mockResolvedValue(customer);
    jest.spyOn(service as any, 'createCustomer');
    // Act
    const result = await service.findOrCreateCustomer({ user });

    // Assert
    expect(result.data).toMatchSnapshot();
    expect((service as any).createCustomer).toHaveBeenCalled();
  });

  it('should find customer', async () => {
    // Arrange
    const customer = {
      id: 'cus_KGjEQgjxBxlXpa',
      object: 'customer',
      name: 'Bogdan Lupu',
    };
    const user = { ...userMock(), stripeCustomerId: customer.id };
    userRepository.findOne = jest.fn().mockResolvedValue(user);
    (service as any).stripeService.customers.list = jest.fn().mockResolvedValue({ data: [customer] });
    userRepository.update = jest.fn().mockResolvedValue(user);

    // Act
    const result = await service.findOrCreateCustomer({ user });

    // Assert
    expect(result.data).toMatchSnapshot();
  });

  xit('should get subscription data', async () => {
    const user = userMock();
    const subscription = stripeSubscription;
    const data = await (service as any).subscriptionData({ user, subscription });
    expect(data).toMatchSnapshot();
  });

  xit('should return Trial Expired subscription', async () => {
    const user = userMock();
    const data = await (service as any).subscriptionData({ user, subscription: {} });
    expect(data).toMatchSnapshot();
  });

  xit('should return Trial subscription', async () => {
    const user = { ...userMock(), createdAt: new Date() };
    const data = await (service as any).subscriptionData({ user, subscription: {} });
    expect(data).toMatchSnapshot();
  });
});
