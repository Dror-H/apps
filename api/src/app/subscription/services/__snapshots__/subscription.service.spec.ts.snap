// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subscription Test suite should create customer 1`] = `
Object {
  "id": "cus_KGjEQgjxBxlXpa",
  "name": "Bogdan Lupu",
  "object": "customer",
}
`;

exports[`Subscription Test suite should find customer 1`] = `
Object {
  "id": "cus_KGjEQgjxBxlXpa",
  "name": "Bogdan Lupu",
  "object": "customer",
}
`;

exports[`Subscription Test suite should get subscription data 1`] = `
Object {
  "active": true,
  "billingCycle": "monthly",
  "cancelAtPeriodEnd": false,
  "color": "#7ed321",
  "id": "sub_1Jf7BWKlYpB2IMP7vWM2pUOh",
  "label": "Active",
  "latestInvoice": "in_1Jf7BWKlYpB2IMP708aJ6Eyd",
  "nextPayment": "01/10/2021",
  "planCost": 19,
  "price": "price_1JewjrKlYpB2IMP7GTT1OKMv",
  "quantity": 1,
  "spendCap": 1,
  "status": "0",
  "used": 100,
}
`;

exports[`Subscription Test suite should return Trial Expired subscription 1`] = `
Object {
  "active": false,
  "billingCycle": "monthly",
  "cancelAtPeriodEnd": undefined,
  "color": "#474b6d",
  "id": undefined,
  "label": "Trial Expired",
  "latestInvoice": undefined,
  "nextPayment": "28/07/2022",
  "planCost": 19,
  "price": undefined,
  "quantity": 1,
  "spendCap": 1,
  "status": "4",
  "used": 100,
}
`;

exports[`Subscription Test suite should return Trial subscription 1`] = `
Object {
  "active": true,
  "billingCycle": "monthly",
  "cancelAtPeriodEnd": undefined,
  "color": "#5f63f2",
  "id": undefined,
  "label": "Trial",
  "latestInvoice": undefined,
  "nextPayment": "28/07/2022",
  "planCost": 19,
  "price": undefined,
  "quantity": 1,
  "spendCap": 1,
  "status": "1",
  "used": 100,
}
`;
